# Results

In [1]:
import pandas as pd

Now that we have our best performing hyperparameter, we can pass in the best param into our model. Then we can train that model on the entire training set and evaluate it on our testing set.

In [3]:
pd.read_csv('../results/finaltable.csv', sep = ";")

Unnamed: 0  mean_fit_time  std_fit_time  param_ridge__alpha  \
0           8       0.018578      0.001218           14.169508   

                               params  mean_test_score  std_test_score  \
0  {'ridge__alpha': 14.1695082925079}        -4.615479        0.298376   

   rank_test_score  final score  
0                1    -3.834295

According to our best model, our RMSE is [3.83] which is better than the value we obtained from cross validation. This suggests one of two things:
1. Our model generalized well on this new dataset, or 
2. We happened to be lucky for the test set and the model wont generalize well to other datasets in practice.

:::{figure-md} relation-fig
<img src="../results/figures/predvsfinal.png" alt="cat" class="bg-primary mb-1" width="400px">

A plot displaying the relation between the predicted grade and the actual grade
:::

The regression graph {numref}`relation-fig` shows that we predict roughly the same values no matter what the true grade of the student happens to be. Only at the very high true grades do we see the predictions start to increase, albeit marginally. The range at low values is a bit larger at about 5-12.5 while at high true grades, it is about 8-13.

Looking at what feature make the most impact in predicting high true grades, we see that the mother has a very large impact on the predicted grades. Depending on the mother's job, it can positively influence the final grade or it can negatively impact it. We also see that being in a romantic relationship seems to be the most impactful at predicting lower grades. Interestingly, study time isn't the most influential part feature despite intuition telling us that with higher study time, we would except to get higher grades.

In [6]:
pd.read_csv('../results/coeff-table.csv', sep = ";").sort_values("coefficients", ascending=False)

Unnamed: 0       features  coefficients
0            1           Medu      0.871048
1            6    Mjob_health      0.783865
2           14   Fjob_teacher      0.684833
3            8  Mjob_services      0.449528
4            0      studytime      0.312556
5            2           Fedu      0.146562
6           11    Fjob_health     -0.066799
7           13  Fjob_services     -0.147888
8           10   Fjob_at_home     -0.209156
9            4     traveltime     -0.229963
10          12     Fjob_other     -0.260990
11           7     Mjob_other     -0.298394
12           9   Mjob_teacher     -0.410360
13          15      Pstatus_T     -0.520461
14           5   Mjob_at_home     -0.524639
15           3          goout     -0.613642
16          16   romantic_yes     -1.244385